File: day-07/src/main.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;
use std::collections::HashMap;

enum  FileSystemObj {
    Dir {
        name: String, 
        subdirs: HashMap<String, Box<FileSystemObj>>,
        files: HashMap<String, i32>, 
        parent: Box<FileSystemObj>},
        Null,
}

fn main() {

    let mut root = FileSystemObj::Dir{
        name: String::from("/"),
        subdirs: HashMap::new(),
        files: HashMap::new(),
        parent: Box::new(FileSystemObj::Null),
    };

    // let mut path: Vec<&FileSystemObj> = Vec::new();
    // path.push(&root);
    let mut cur_dir = &mut root;

    if let Ok(lines) = read_lines("./input.txt") {
        for line in lines {
            if let Ok(cmd) = line {
                let tokens: Vec<&str> = cmd.split(' ').collect();

                //Look at first letter
                match tokens[0] {
                    // Commands
                    "$" => {
                        //println!{"Command: {}", cmd};
                        //Only expecting cd or ls...
                        match tokens[1] {
                            "cd" => {
                                //println!("cd to {}", tokens[2]);
                                match tokens[2] {
                                    ".." => {
                                        if let FileSystemObj::Dir{name:_, subdirs:_, files:_,  parent} = cur_dir {
                                            cur_dir = parent;
                                        }
                                    },
                                    _ => {
                                        // cd to a new dir. Let's assume that it's been seen before
                                        // (leveraging knowledge about the input file)

                                        if let FileSystemObj::Dir{name:_, subdirs, files:_, parent:_} = cur_dir {
                                            cur_dir = &mut subdirs.get_mut(tokens[3]).unwrap();
                                        }
                                    }
                                }
                            },
                            "ls" => {/*Can skip ls, data follows...*/},
                            _ => {println!("Unexpected cmd: {}", tokens[1]);}
                        }
                    },
                    // Directory or File data 
                    _ => {
                        println!{"Data: {}", cmd}
                        match tokens[0] {
                            "dir" => {
                            },
                            _ => { //Must be a file, let's add it to the cur_dir's HashMap of files
                                if let FileSystemObj::Dir{name:_, subdirs:_, ref mut files, parent:_} = cur_dir {
                                    files.insert(String::from(tokens[1]),tokens[0].parse().unwrap());
                                }
                            }
                        }
                    }
                }        
            }
        }
    }
}

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where P: AsRef<Path>, {
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

File: day-07/main.py

global_accumulator = 0

# Part 2 Requirements:
#   Disk space: 70000000 
#   Currenlty used: 47442399
#   Required Min: 30000000 
global_min_size = 30000000
global_available_space = (70000000 - 47442399)

class Dir:

    def __init__(self, name,):
        self.name = name
        self.sub_dirs = dict()
        self.files = dict()
    
    def __str__(self):
        return f'Name: {self.name}, dirs: {len(self.sub_dirs)}, files: {len(self.files)}'

    def get_total_size(self):
        total_size = 0
        # Add my file sizes
        for file_size in self.files.values():
            total_size += file_size
        # Add my sub_dir sizes
        for d in self.sub_dirs.values():
            total_size += d.get_total_size()

        #print(f'Total size of {self.name} is {total_size}')

        if total_size <= 100000:
            global global_accumulator
            global_accumulator += total_size

        global global_min_size
        global global_avaialable_space
    
        if global_available_space + total_size >= 30000000 and total_size < global_min_size:
            global_min_size = total_size
            print(f"New target dir: {total_size}")

        return total_size

root = Dir('/')
cur_dir = root

f = open('input.txt', 'r', encoding='utf-8')

for line in f:
    tokens = line.split()

    # Is it a command? We care about "cd .." and "cd dir_name". We can skip "ls".
    if tokens[0] == '$':
        if tokens[1] == 'cd':
            if tokens[2] == '..': # cd'ing up to parent_dir
                cur_dir = cur_dir.parent_dir
            else: # cd'ing to a sub_dir
                cur_dir = cur_dir.sub_dirs[tokens[2]]
    # Or a dir entry
    elif tokens[0] == 'dir':
        temp_dir = Dir(tokens[1])
        temp_dir.parent_dir = cur_dir
        cur_dir.sub_dirs[tokens[1]] = temp_dir
    else: #it's a file entry
        cur_dir.files[tokens[1]] = int(tokens[0]) # nnnnnn filename 

total_size = 0
total_size += root.get_total_size()
print(f'Total size: {total_size}')
print(f'Root Info: {root}')
print(f'Global Accumulator: {global_accumulator}')


